Snap 589

dshirish opened the pull request:
Snap-589: A procedure to check index consistency
1. In order to run existing derby proc CHECK_TABLE procedure on every node , added a wrapper procedure called SYS.CHECK_TABLE_EX(String schema, String table) in GfxdSystemProcedures. This will send a GfxdSystemProcedureMessage to all data nodes where CHECK_TABLE will be executed locally.

User will call it as 'values SYS.CHECK_TABLE_EX('schema', 'table')'. If the indexes are consistent, it just returns a value '1' otherwise it will throw an exception.
1. Tested for local index by deleting/ corrupting entries in the skipList and making sure that the inconsistency is detected. Added  tests in CheckTableDUnit

Currently looking into optimizing already present implementation for global index consistency check. Opening the PR for current set of changes of local index in the mean time.

@kneeraj @sumwale , please review.

dshirish committed:
Changes to get derby check_table proc execute on all nodes
Added a unit tests for checking consistency of local index using check_table

dshirish committed:
Merge branch 'snappy/master' into SNAP-589

dshirish committed:
Removed debug statements

dshirish committed:
Mionr modifications to the test

ymahajan commented:
Will i know the inconsistent row(s) from the exception message ? May be we should add a procedure to repair the index itself, it may be a good option rather than dropping the index.

dshirish commented:
Yes, for example the message is of the following format if value is corrupted-

"The values for column 'COL2' in index 'IDX2' and table 'TEST.TABLE1' do not match for row location VMBucketRowLocationThinRegionEntryHeap@13933f1e(key=100; rawValue=(2,0,0,0,99,0,0,0,99,57,57,1,5,9); lockState=0x0; bucketId=2).  The value in the index is '0', while the value in the base table is '99'.  The full index key, including the row location, is '{ 0, VMBucketRowLocationThinRegionEntryHeap@13933f1e(key=100; rawValue=(2,0,0,0,99,0,0,0,99,57,57,1,5,9); lockState=0x0; bucketId=2) }'.  The suggested corrective action is to recreate the index."

It will abort immediately on first failure. If there # of entries mismatch also it will abort.

dshirish committed:
Optimized the consistency checks for global index.  Now a base table is scanned and existence of keys is checked using getAll onl global index region in a batch of 1000.
Also making sure that the size of base table and global index region is same using region.size.

Added unit tests

dshirish committed:
Merge branch 'snappy/master' into SNAP-589

dshirish committed:
Return true in GfxdSystemProcedureMessage.SysProcMethod#initConnection for GfxdSystemProcedureMessage.SysProcMethod#checkTableEx so that lcc is created

dshirish committed:
Return true in GfxdSystemProcedureMessage.SysProcMethod#initConnection for GfxdSystemProcedureMessage.SysProcMethod#checkTableEx so that lcc is created

dshirish committed:
Verify global index size on one of the data stor nodes instead of all

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/catalog/GfxdSystemProcedures.java:
```diff
@@ -2014,6 +2014,33 @@ public static void CANCEL_STATEMENT(String statementUUID)
+    try {
+      if (GemFireXDUtils.getMyVMKind().isStore()) {
+        GfxdSystemProcedureMessage.SysProcMethod.checkTableEx.processMessage(params,
+            Misc.getMyId());
```
We should invoke check in parallel on all the nodes including this store vm. Since it is a time taking op so it is all the more important. Probably just including this vm id in the send target inside publishMessage.

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/engine/ddl/catalog/messages/GfxdSystemProcedureMessage.java:
```diff
@@ -1473,6 +1478,95 @@ String getSQLStatement(Object[] params) throws StandardException {
+        String table = (String)params[1];
+        // member on which global index size are verified
+        DistributedMember m = (DistributedMember)params[2];
+        SanityManager.DEBUG_PRINT(GfxdConstants.TRACE_SYS_PROCEDURES,
```
Just review the local var names, a bit more descriptive.

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/iapi/db/ConsistencyChecker.java:
```diff
@@ -148,11 +154,20 @@ public static boolean checkTable(String schemaName, String tableName)
 		SchemaDescriptor		sd;
 		ConstraintDescriptor	constraintDesc;
 
+		LocalRegion region = (LocalRegion)Misc.getRegionForTable(schemaName + "." + tableName, true);
```
May be a null check for schema here or upper level.

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/iapi/db/ConsistencyChecker.java:
```diff
@@ -504,170 +530,267 @@ public static boolean checkTable(String schemaName, String tableName)
+		for (Object o : tableIndexKeys) {
+			if (globalIndexEntriesMap.get(o) == null) {
+				DataValueDescriptor keyColumns[] = new DataValueDescriptor[searchCondition.length];
+				if (searchCondition.length == 1) {
```
May be later we can add routing object correctness with the actual column's routing object using the partitoin resolver. We should also have a ticket to trac a lighter overall check for index consistency which will just check the sizes and come back.

kneeraj commented on gemfirexd/core/src/main/java/com/pivotal/gemfirexd/internal/iapi/db/ConsistencyChecker.java:
```diff
@@ -504,170 +530,267 @@ public static boolean checkTable(String schemaName, String tableName)
+				key = searchCondition[0].getClone();
+			} else {
+				key = new CompositeRegionKey(getNewArray(searchCondition,
+						baseColumns, true));
```
Since we are using a fixed batch of size 1000 then we can just have 1000 empty object created upfront and reuse the same again and again. Just using some setter methods.

dshirish committed:
Allocated CRK in a fixed array size of 5000 upfront. Incorporates review comment https://github.com/SnappyDataInc/snappy-store/pull/38#discussion_r57726068
Also added a unit test for composite key

dshirish committed:
Incorporate minor review comments

dshirish committed:
Merge branch 'snappy/master' into SNAP-589

dshirish committed:
Incorporate review comments. Executing the proc in a new thread self and in parallel sending message to execute it on ther nodes.

dshirish commented:
Incorporated all review comments, also added a task (SNAP-654) to add a lighter proc that will just verify index sizes.
